### PR TITLE
[BEAM-2969] Handle negative AVRO timestamps

### PR DESCRIPTION
**Pull request by udim**
from BigQuery batch loads.

Also simplifies the code a bit. Microsecond part will have trailing
zeroes.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
R: @chamikaramj

**Comment by udim** [comment]
run java postcommit

**Comment by udim** [comment]
R: @pabloem as well

**Comment by udim** [comment]
Lots of GCS 503s. Retrying tests one-by-one.

**Comment by udim** [comment]
Run Java PreCommit

**Comment by udim** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by udim** [comment]
Run JavaPortabilityApi PreCommit

**Comment by udim** [comment]
run java postcommit

**Comment by pabloem** [comment]
Woah this is pretty tricky. I was wondering yesterday what happens with leap years, leap seconds, and daylight savings with these things...
It looks good to me. I'll take another look, and wait for the tests to pass.

**Comment by pabloem** [comment]
What happens if we need 64 bits to represent a positive time? Maybe that's too far in the future?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java [hunk 1]
```diff
@@ -311,12 +304,10 @@ private static Object convertRequiredField(
...
       case "TIMESTAMP":
-        // TIMESTAMP data types are represented as Avro LONG types. They are converted back to
-        // Strings with variable precision (up to six digits) to match the JSON files exported by
-        // BigQuery.
+        // TIMESTAMP data types are represented as Avro LONG types, microseconds since the epoch.
+        // Values may be negative since BigQuery timestamps start at 0001-01-01 00:00:00 UTC.
         verify(v instanceof Long, "Expected Long, got %s", v.getClass());
-        double doubleValue = ((Long) v) / 1_000_000.0;
-        return formatTimestamp(Double.toString(doubleValue));
+        return formatTimestamp((Long) v);
```
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java]: Maybe you don't need to cast it if you've checked its type?

**Comment by pabloem** [comment]
This LGTM. Just hoping for an answer to my previous question, and I'll merge after tests pass. Thanks!

**Comment by udim** [comment]
The timestamp we get from BQ is a signed 64-bit integer.
```py
>>> (2**63-1) / 1000000. / 86400 / 365
292471.208677536
```
We have about 290,000 years left. :)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java (hunk 1 above)
Comment (udim) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java]: I had to cast that explicitly; `v` is of type `Object`.

**Comment by pabloem** [comment]
Thanks Udi

**Pull request merged**